File: app/src/main/java/de/htw/berlin/s0558606/lasersensorcommunicator/io/CSVWriter.java
package de.htw.berlin.s0558606.lasersensorcommunicator.io;

import android.app.ProgressDialog;
import android.content.Context;
import android.database.Cursor;
import android.os.Environment;
import android.os.Handler;
import android.os.Message;

import java.io.File;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.List;

import de.htw.berlin.s0558606.lasersensorcommunicator.R;
import de.htw.berlin.s0558606.lasersensorcommunicator.model.AppDatabase;
import de.htw.berlin.s0558606.lasersensorcommunicator.model.Location;
import de.htw.berlin.s0558606.lasersensorcommunicator.model.Measurement;
import de.htw.berlin.s0558606.lasersensorcommunicator.model.SensorData;

/**
 * Created by Marcel Ebert S0558606 on 08.06.18.
 */
public class CSVWriter {

    public static String getDatabaseContentAsCSV(Context context) {

        File folder = new File(Environment.getExternalStorageDirectory() + "/Folder");

        boolean var = false;
        if (!folder.exists())
            var = folder.mkdir();

        System.out.println("Folder created = " + var);

        final String fileName = folder.toString() + "/" + "sensordata.csv";

        createFileFromLists(getListOfDatabaseObjects(context), fileName);

        return fileName;
    }

    private static List<List<String>> getListOfDatabaseObjects(Context context) {
        AppDatabase db = AppDatabase.getInstance(context);

        List<List<String>> superList = new ArrayList<>();

        List<Location> locations = db.locationDao().getAllLocationsSynchronous();

        for (Location location : locations) {
            List<Measurement> measurements = db.measurementDao().findMeasurementsByLocationIDSynchronous(location.getId());

            for (Measurement measurement : measurements) {
                List<SensorData> sensorDataList = db.sensorDataDao().findDataByMeasurementIdSynchronous(measurement.getId());

                for (SensorData sensorData : sensorDataList) {
                    List<String> miniList = new ArrayList<>();
                    miniList.add(sensorData.getDateAsString());
                    miniList.add(sensorData.getPm10());
                    miniList.add(sensorData.getPm25());
                    miniList.add(String.valueOf(location.getLocation().latitude));
                    miniList.add(String.valueOf(location.getLocation().longitude));
                    superList.add(miniList);
                }
            }
        }
        return superList;

    }

    private static void createFileFromLists(List<List<String>> superList, String fileName) {
        try {
            FileWriter fw = new FileWriter(fileName);

            fw.append("Timestamp");
            fw.append(',');

            fw.append("PM10");
            fw.append(',');

            fw.append("PM25");
            fw.append(',');

            fw.append("Latitude");
            fw.append(',');

            fw.append("Longitude");

            fw.append('\n');

            for (List<String> miniList : superList) {
                for (String value : miniList) {
                    fw.append(value);
                    fw.append(",");
                }
            }

            // fw.flush();
            fw.close();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: app/src/main/java/de/htw/berlin/s0558606/lasersensorcommunicator/io/CSVWriter.kt
package de.htw.berlin.s0558606.lasersensorcommunicator.io

import android.content.Context
import de.htw.berlin.s0558606.lasersensorcommunicator.model.AppDatabase
import java.io.File
import java.io.FileWriter
import java.util.*

/**
 * Created by Marcel Ebert S0558606 on 08.06.18.
 */
object CSVWriter {

    fun getDatabaseContentCSVFileName(context: Context): String {

        val folderPath = getFolderPath(context)
        val fileName = "$folderPath/sensordata.csv"

        val databaseContentLists = getListOfDatabaseObjects(context)

        tryWriteFileFromLists(databaseContentLists, fileName)

        return fileName
    }

    private fun getFolderPath(context: Context): String {
        val folder = File(context.filesDir.toString() + "/SensorData")

        if (!folder.exists())
            folder.mkdir()

        return folder.toString()
    }

    private fun getListOfDatabaseObjects(context: Context): List<List<String>> {
        val db = AppDatabase.getInstance(context)

        val superList = ArrayList<List<String>>()

        val locations = db.locationDao().getAllLocationsSynchronous()

        for (location in locations) {
            val measurements = db.measurementDao().findMeasurementsByLocationIDSynchronous(location.id)

            for (measurement in measurements) {
                val sensorDataList = db.sensorDataDao().findDataByMeasurementIdSynchronous(measurement.id)

                for (sensorData in sensorDataList) {
                    val miniList = ArrayList<String>()
                    miniList.add(sensorData.getDateAsString())
                    miniList.add(sensorData.pm10)
                    miniList.add(sensorData.pm25)
                    miniList.add(location.location.latitude.toString())
                    miniList.add(location.location.longitude.toString())
                    superList.add(miniList)
                }
            }
        }
        return superList

    }

    private fun tryWriteFileFromLists(superList: List<List<String>>, fileName: String) {
        try {
            val fw = FileWriter(fileName)

            fw.append("Timestamp")
            fw.append(',')

            fw.append("PM10")
            fw.append(',')

            fw.append("PM25")
            fw.append(',')

            fw.append("Latitude")
            fw.append(',')

            fw.append("Longitude")

            fw.append('\n')

            for (miniList in superList) {
                miniList.forEachIndexed { index, value ->
                    run {
                        if (index != miniList.lastIndex) {
                            fw.append(value)
                            fw.append(",")
                        } else {
                            fw.append(value)
                        }
                    }
                }

                fw.append("\n")
            }

            fw.close()

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

}
